src/Chapter3.hs

```diff
@@ -1006,6 +1114,130 @@ Implement data types and typeclasses, describing such a battle between two
+
+  getActions :: a -> [Action]
+
+data Knight = Knight {
+  knightHealth :: Health,
+  knigthAttack :: Attack,
+  knightDefence :: Defence,
+  knightActions :: [Action]
+} deriving (Show)
+
+instance Fighter Knight where
+  getHealth :: Knight -> Health
+  getHealth knight = knightHealth knight
+
+  getAttack :: Knight -> Attack
+  getAttack knight = knigthAttack knight
+
+  setHealth :: Health -> Knight -> Knight
+  setHealth health knight = knight { knightHealth = health }
+
+  getDefence :: Knight -> Defence
+  getDefence knight = knightDefence knight
+
+  setDefence :: Defence -> Knight -> Knight
+  setDefence defence knight = knight { knightDefence = defence }
+
+  getActions :: Knight -> [Action]
+  getActions knight = knightActions knight
+
+
+drinkPotion :: (Fighter a) => Potion -> a -> a
+drinkPotion (Potion recoverHealth) knight = setHealth (Health newHealth) knight
+  where
+    Health health = getHealth knight
+    newHealth = health + recoverHealth
+
+castSpell :: (Fighter a) => Spell -> a -> a
+castSpell (Spell incDefence) knight = setDefence (Defence newDefence) knight
+  where
+    Defence defence = getDefence knight
+    newDefence = defence + incDefence
+
+
+data Monster = Monster {
+  monsterHealth :: Health,
+  monsterAttack :: Attack,
+  monsterActions :: [Action]
+} deriving (Show)
+
+instance Fighter Monster where
+  getHealth :: Monster -> Health
+  getHealth monster = monsterHealth monster
+
+  getAttack :: Monster -> Attack
+  getAttack monster = monsterAttack monster
+
+  setHealth :: Health -> Monster -> Monster
+  setHealth health monster = monster { monsterHealth = health }
+
+  getActions :: Monster -> [Action]
+  getActions monster = monsterActions monster
+
+calculateHealth :: Attack -> Defence -> Health -> Health
+calculateHealth (Attack damage) (Defence defence) (Health health) = Health $ health - max 0 (damage - defence)
+
+
+fight :: (Fighter a, Fighter b) => a -> b -> Result
+fight firstFighter secondFighter = go SecondFighterWin firstFighter secondFighter combineActions
+  where
+    combineActions :: [Action]
+    combineActions = concat [fa:sa:[] | (fa, sa) <- zip (cycle $ getActions firstFighter) (cycle $ getActions secondFighter)]
+    
+    go :: (Fighter a, Fighter b) => Result -> a -> b -> [Action] -> Result
+    go result activeFighter passiveFighter (action:actions) 
+      | isDead activeFighter = result
+      | otherwise = case action of
+                      AttackAction      
+                        -> go (nextResult result) (attack activeFighter passiveFighter) activeFighter actions
+                      DrinkPotionAction (Potion potion)
+                        -> go (nextResult result) passiveFighter (drinkPotion (Potion potion) activeFighter) actions
+                      CastSpellAction (Spell spell)
+                        -> go (nextResult result) passiveFighter (castSpell (Spell spell) activeFighter) actions
+                      RunAwayAction
+                        -> result
```

Comment:
Interesting concept of swapping fighters with each other at each iteration and use `Bound` instance of `Result` in here 🧠 

One moment, in your case it is actually there is no difference between Monster and Knight, and you could use one type for that.
In the task description we advice two types, as we suggest separate actions for Knight and Monster, so that the Monster can't heal, and for example, knight can't run away.

But your interpretation is very interesting, and I think that you brilliantly implemented all functions 👍🏼 Weel done!